StripeFinancialConnections/StripeFinancialConnections/Source/Native/PartnerAuth/PartnerAuthViewController.swift

```diff
@@ -235,24 +235,14 @@ final class PartnerAuthViewController: SheetViewController {
                 completionHandler: { [weak self] isSuccess in
                     guard let self = self else { return }
                     if !isSuccess {
-                        self.dataSource.recordAuthSessionEvent(
-                            eventName: "cancel",
-                            authSessionId: authSession.id
-                        )
-
-                        // cancel current auth session
-                        self.dataSource.cancelPendingAuthSessionIfNeeded()
-
-                        // whether legacy or OAuth, we always go back
-                        // if we got an explicit cancel from backend
-                        self.navigateBack()
+                        handleAuthSessionCancel(authSession, nil)
```

Comment:
We can see the previous logic on the left side.

Reply:
The new code is:

```
    private func handleAuthSessionCancel(
        _ authSession: FinancialConnectionsAuthSession,
        _ error: Error?
    ) {
        if authSession.isOauthNonOptional {
            // on "manual cancels" (for OAuth) we log retry event:
            dataSource.recordAuthSessionEvent(
                eventName: "retry",
                authSessionId: authSession.id
            )
        } else {
            // on "manual cancels" (for Legacy) we log cancel event:
            dataSource.recordAuthSessionEvent(
                eventName: "cancel",
                authSessionId: authSession.id
            )
        }
        // cancel current auth session because something went wrong
        dataSource.cancelPendingAuthSessionIfNeeded()
        if authSession.isOauthNonOptional {
            // for OAuth institutions, we remain on the pre-pane,
            // but create a brand new auth session
            createAuthSession()
        } else {
            // for legacy (non-OAuth) institutions, we navigate back to InstitutionPickerViewController
            navigateBack()
        }
    }
```

Reply:
so we can see that there's only two additions, and they are both for `isOauthNonOptional`:

we add:

```
            // on "manual cancels" (for OAuth) we log retry event:
            dataSource.recordAuthSessionEvent(
                eventName: "retry",
                authSessionId: authSession.id
```

and

```
            // for OAuth institutions, we remain on the pre-pane,
            // but create a brand new auth session
            createAuthSession()
```